include/ua_securitypolicy.h

```diff
@@ -0,0 +1,235 @@
+
+typedef struct
+{
+    /**
+     * \brief Encrypts the given plaintext using a symmetric algorithm and key.
+     *
+     * \param securityContext the SecurityContext to work on.
+     * \param data the data to encrypt. The data will be encrypted in place.
+     *             The implementation may allocate additional memory though.
+     */
+    UA_StatusCode (*const encrypt)(const UA_Channel_SecurityContext* const securityContext,
+                                   UA_ByteString* const data);
+
+    /**
+     * \brief Decrypts the given ciphertext using a symmetric algorithm and key.
+     *
+     * \param securityContext the SecurityContext which contains information about the keys needed to decrypt the message.
+     * \param data the data to decrypt. The decryption is done in place.
+     */
+    UA_StatusCode (*const decrypt)(const UA_Channel_SecurityContext* const securityContext,
+                                   UA_ByteString* const data);
+
+    /**
+     * \brief Pseudo random function that is used to generate the symmetric keys.
+     *
+     * For information on what parameters this function receives in what situation,
+     * refer to the OPC UA specification 1.03 Part6 Table 33
+     *
+     * \param secret
+     * \param seed
+     * \param out an output to write the data to. The length defines the maximum
+     *            number of output bytes that are produced.
+     */
+    UA_StatusCode (*const generateKey)(const UA_ByteString* const secret,
+                                       const UA_ByteString* const seed,
+                                       UA_ByteString* const out);
+    /**
+     * \brief Random generator for generating nonces.
+     * 
+     * Generates a nonce. The length will be the same as the symmetric encryptingKeyLength
+     * as specified in the OPC UA Specification Part 4 - Services 1.03 on page 20
+     * 
+     * \param securityPolicy the securityPolicy this function is invoked on. Example: myPolicy->generateNonce(myPolicy, &outBuff);
+     * \param out pointer to a buffer to store the nonce in. Needs to be allocated by the caller.
+     *            The size must be equal to the symmetric encrypting key length.
+     */
+    UA_StatusCode (*const generateNonce)(const UA_SecurityPolicy* const securityPolicy,
+                                         UA_ByteString* const out);
+
+    /**
+     * \brief Calculates the padding size for a message with the specified amount of bytes.
+     *
+     * \param securityPolicy the securityPolicy the function is invoked on.
+     * \param bytesToWrite the size of the payload plus the sequence header, since both need to be encoded
+     * \param paddingSize out parameter. Will contain the paddingSize byte.
+     * \param extraPaddingSize out parameter. Will contain the extraPaddingSize. If no extra padding is needed, this is 0.
+     */
+    UA_StatusCode (*const calculatePadding)(const UA_SecurityPolicy* const securityPolicy,
```

Comment:
This could just return paddingSize instead of UA_StatusCode....
I don't see a situation where an error case is necessary here.